### PR TITLE
Fixed #29490 -- Added support for object-based Media CSS and JS paths.

### PR DESCRIPTION
**Pull request by claudep**

### CODE REVIEW

File: tests/forms_tests/tests/test_media.py [hunk 1]
```diff
@@ -646,3 +648,113 @@ def test_add_empty(self):
...
+@html_safe
+class Asset:
+    def __init__(self, path):
+        self.path = path
+
+    def __eq__(self, other):
+        return (
+            (self.__class__ == other.__class__ and self.path == other.path) or
+            (other.__class__ == str and self.path == other)
+        )
```
Comment (kezabelle) [inline tests/forms_tests/tests/test_media.py:662]: Is this only required for the specifics of the test cases, or is it required to satisfy the merging and sorting in `Media` itself? If the former, all well and good, if the latter, would that need to be part of the `Paths as objects` interface contract?

Comment (claudep) [reply tests/forms_tests/tests/test_media.py:662]: Well spotted. This should probably be added to the interface contract. The more we add to that contract, the more this pleads for a base asset class proposed by Django, but I'd still like this PR as simplest as possible on the Django code side.

**Comment by interDist** [comment]
Claude, what is your reason for choosing to require users to supply a whole class incl. their own implementation of the corresponding html tag? Wouldn't be allowing usage of dictionaries in lieu of strings much easier for the users?

Then a switch would be trivial: 

    js = ("not-my-script.js", "another-script.js")

Becomes

    js = ({"src": "not-my-script.js", "integrity": "...", "defer": True}, "another-script.js")

**Comment by claudep** [comment]
The main reason is flexibility. In my experience, I often started by using dictionaries in such cases, and soon I needed some custom logic for which a class was a better abstraction. We could also consider supporting dicts like you suggest, and then the contract may be to provide an dict/object with all `keys()` being the tag attributes. But then building the final html tag would continue to be the task of the Media class, and if any need arises to change the way the html is produced, we are stuck. It can be compared to the Django form layer, when we started with some hard-coded html construction, to reach a point where now every component is an object with a `render()` method. In my opinion, this approach is the most flexible in the long term. This is open for discussion, of course.

**Comment by interDist** [comment]
>  if any need arises to change the way the html is produced

It's a bit hard to imagine what revolutionary can happen with the CSS & JS tags 🙃  They have an "src" attribute and a bunch of other attributes, which continue to evolve over time but still remain just attributes (either boolean ones or with a value)...

**Comment by claudep** [comment]
> It's a bit hard to imagine what revolutionary can happen with the CSS & JS tags upside_down_face They have an "src" attribute and a bunch of other attributes, which continue to evolve over time but still remain just attributes (either boolean ones or with a value)...

One example (maybe outdated) is someone wanting to produce some XHTML-compatible tag (ending with `/>`).
Another example (maybe unrealistic) is for an absolute path, someone may want to download the resource to some local static directory, and then serve it locally.
Let's not debate these exact examples in details, they only show what `kind` of cleverness one may want to add to asset paths.

**Comment by interDist** [comment]
I see. Would it be an idea then to provide some base class(es) as you did for `FormsMediaObjectTestCase`, for the sake of those who only want to inject an extra attribute into the js/css tag?

**Review by matthiask — APPROVED** [review]
I like it and I like the fact that this would probably make [django-js-asset](https://github.com/matthiask/django-js-asset) obsolete. 

No comments on the implementation except that everything seems clear and well thought out. I especially like that `JS` and `CSS` objects can replace the full representation of the tag and not just the path element and a few hardcoded extra attributes.

Thanks!

**Comment by claudep** [comment]
Thanks Matthias. Do you think that the base Asset/JS/CCS classes currently in test files only should be provided as base utility classes in Django core?

**Comment by matthiask** [comment]
I don't think that has to be done right now. The `JS` class in the testsuite only supports the `integrity` attribute while django-js-asset's `JS` class supports arbitrary attributes. So contrary to what I wrote yesterday it wouldn't replace django-js-asset but it certainly would make its implementation less fragile. Not every battery has to be included in Django.

**Comment by felixxm** [comment]
@claudep Thanks :+1:

I rebased from the `main` branch, applied `black`, and pushed small edits to docs.

**Review by carltongibson — APPROVED** [review]
Very nice. Thanks @claudep! 🎁

File: docs/topics/forms/media.txt [hunk 2]
```diff
@@ -254,6 +257,28 @@ Or if :mod:`~django.contrib.staticfiles` is configured using the
...
 
+Paths as objects
+----------------
+
+.. versionadded:: 4.1
+
+Asset paths may also be given as hashable objects implementing an
+``__html__()`` method. The ``__html__()`` method is typically added using the
+:func:`~django.utils.html.html_safe` decorator. The object is responsible to
+output the complete HTML ``<script>`` or ``<link>`` tag content::
```
Comment (carltongibson) [inline docs/topics/forms/media.txt:268]:
```suggestion
:func:`~django.utils.html.html_safe` decorator. The object is responsible for
outputting the complete HTML ``<script>`` or ``<link>`` tag content::
```

**Pull request merged**